File: mongo/JavaMongo/src/UpdateOne.java
// db.collection.updateOne() method example in Java

import com.mongodb.client.*;
import com.mongodb.client.result.UpdateResult;
import org.bson.Document;
import org.bson.types.ObjectId;

import java.util.Arrays;

public class UpdateOne {
    public static void main(String[] args) {
        String uri = "mongodb://bigboxuser:bigboxpass@localhost:27017/";

        try (MongoClient mongoClient = MongoClients.create(uri)) {
            MongoDatabase database = mongoClient.getDatabase("bigboxcode");
            MongoCollection<Document> customerCollection = database.getCollection("customer");

            // Insert some documents
            Document doc1 = new Document("name", "Izabella Conroy")
                    .append("age", 19)
                    .append("phone", "108-679-2154")
                    .append("address", new Document("street", "11 North Vineyard Drive")
                            .append("city", "Pueblo West")
                            .append("state", "CO")
                            .append("postalCode", "81007"));
            Document doc2 = new Document("name", "Alisa Parker")
                    .append("age", 72)
                    .append("phone", "768-319-1054")
                    .append("address", new Document("street", "8532 Ingalls Circle")
                            .append("city", "Arvada")
                            .append("state", "CO")
                            .append("postalCode", "80003"));
            Document doc3 = new Document("name", "Eloise Weber")
                    .append("age", 29)
                    .append("phone", "618-357-2104")
                    .append("address", new Document("street", "632 Belmar Drive")
                            .append("city", "Edmond")
                            .append("state", "OK")
                            .append("postalCode", "73025"));
            customerCollection.insertMany(Arrays.asList(doc1, doc2, doc3));

            // Check the changed document
            ObjectId objId = (ObjectId) doc1.get("_id");
            Document result = customerCollection.find(new Document("_id", objId)).first();
            System.out.println("Command: db.customer.find(objId) | Result: " + result);

            // Select by id and change name, age, and city
            UpdateResult updateResult = customerCollection.updateOne(
                    new Document("_id", objId),
                    new Document("$set", new Document("name", "My Name Changed 9").append("age", 100).append("address.city", "city88"))
            );
            System.out.println("Command: db.customer.updateOne(......find by id, ......update name, age, city) | Result: " + updateResult);

            // Select by id. Change state. Increase age by 7
            updateResult = customerCollection.updateOne(
                    new Document("_id", objId),
                    new Document("$set", new Document("address.state", "ABC")).append("$inc", new Document("age", 7))
            );
            System.out.println("Command: db.customer.updateOne(......find by id, ......update state, age) | Result: " + updateResult);

            // Filter by name. Change age
            updateResult = customerCollection.updateOne(
                    new Document("name", "Alisa Parker"),
                    new Document("$set", new Document("age", 10))
            );
            System.out.println("Command: db.customer.updateOne(......find by name, ......update age) | Result: " + updateResult);

            // Filter by non-existing name. Try to change age
            updateResult = customerCollection.updateOne(
                    new Document("name", "Non Existing Name"),
                    new Document("$set", new Document("age", 100))
            );
            System.out.println("Command: db.customer.updateOne(......find by non existing name, ......update age) | Result: " + updateResult);

            // Filter by non-existing name. Update age. Provide upsert=true. New document created
            updateResult = customerCollection.updateOne(
                    new Document("name", "Non Existing Name"),
                    new Document("$set", new Document("age", 100)),
                    new com.mongodb.client.model.UpdateOptions().upsert(true)
            );
            System.out.println("Command: db.customer.updateOne(......find by non existing name, ......update age, .....provide upsert option) | Result: " + updateResult);

            // Check upserted document
            result = customerCollection.find(new Document("name", "Non Existing Name")).first();
            System.out.println("Command: db.customer.find({ \"name\":  \"Non Existing Name\" }) | Result: " + result);

            // Close the connection
            mongoClient.close();
        }
    }
}

File: mongo/go/updateone/updateone.go
// MongoDB db.collection.updateOne() method examples in Golang

package main

import (
	"context"
	"fmt"
	"os"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	_ "go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

const uri = "mongodb://bigboxuser:bigboxpass@localhost:27017/?maxPoolSize=20&w=majority"

type Customer struct {
	ID      int     `bson:"_id,omitempty"`
	Name    string  `bson:"name"`
	Age     int     `bson:"age"`
	Phone   string  `bson:"phone,omitempty"`
	Address Address `bson:"address,omitempty"`
}

type Address struct {
	Street     string `bson:"street,omitempty"`
	City       string `bson:"city,omitempty"`
	State      string `bson:"state,omitempty"`
	PostalCode string `bson:"postalCode,omitempty"`
	Country    string `bson:"country,omitempty"`
}

func main() {
	// Initiate mongo client
	mongoClient, err := mongo.Connect(context.TODO(), options.Client().ApplyURI(uri))
	if err != nil {
		fmt.Println("MongoDB connection error: ", err)
		os.Exit(1)
	}
	defer mongoClient.Disconnect(context.Background())

	// Initiate collection instance
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	customerCollection := mongoClient.Database("bigboxcode").Collection("customer")

	// Insert some documents
	customers := []interface{}{
		Customer{
			Name:  "Izabella Conroy",
			Age:   19,
			Phone: "108-679-2154",
			Address: Address{
				Street:     "11 North Vineyard Drive",
				City:       "Pueblo West",
				State:      "CO",
				PostalCode: "81007",
			},
		},
		Customer{
			Name:  "Alisa Parker",
			Age:   72,
			Phone: "768-319-1054",
			Address: Address{
				Street:     "8532 Ingalls Circle",
				City:       "Arvada",
				State:      "CO",
				PostalCode: "80003",
			},
		},
		Customer{
			Name:  "Eloise Weber",
			Age:   29,
			Phone: "618-357-2104",
			Address: Address{
				Street:     "632 Belmar Drive",
				City:       "Edmond",
				State:      "OK",
				PostalCode: "73025",
			},
		},
	}

	insertResult, err := customerCollection.InsertMany(ctx, customers)

	if err != nil {
		fmt.Println("Command: db.customer.insertMany(... 3 documents ) | ERROR: ", err)
	}

	fmt.Println("Command: db.customer.insertMany(... 3 documents ) | Result: ", insertResult)

	//  Select by id and change name, age and city
	objectId := insertResult.InsertedIDs[0] // primitive.ObjectIDFromHex("65d3219582913eadf267bbc2")
	filter := bson.M{"_id": objectId}
	update := bson.M{"$set": bson.M{"name": "My Name Changed 9", "age": 100, "address.city": "city88"}}

	// Perform updateOne operation
	updateResult, err := customerCollection.UpdateOne(ctx, filter, update)
	if err != nil {
		fmt.Println("Command: db.customer.updateOne(......find by id, ......update name, age, city) | Error: ", err)
	}

	fmt.Println("Command: db.customer.updateOne(......find by id, ......update name, age, city) | Result: ", updateResult)

	// Check the changed document
	var customer bson.M
	err = customerCollection.FindOne(ctx, filter).Decode(&customer)
	if err != nil {
		fmt.Println("db.customer.find(ObjectId('65d3219582913eadf267bbc2')) | Error: ", err)
	}

	fmt.Println("db.customer.find(ObjectId('65d3219582913eadf267bbc2')) | Result: ", customer)

	// Update document with specified ObjectId, changing state and increasing age
	update = bson.M{"$set": bson.M{"address.state": "ABC"}, "$inc": bson.M{"age": 7}}
	updateResult, err = customerCollection.UpdateOne(ctx, filter, update)
	if err != nil {
		fmt.Println("Command: db.customer.updateOne(......find by id, ......update state, age) | Error: ", err)
	}

	fmt.Println("Command: db.customer.updateOne(......find by id, ......update state, age) | Result: ", updateResult)

	// Update document by filtering with name
	filter = bson.M{"name": "Alisa Parker"}
	update = bson.M{"$set": bson.M{"age": 10}}
	updateResult, err = customerCollection.UpdateOne(ctx, filter, update)
	if err != nil {
		fmt.Println("Command: db.customer.updateOne(......find by name, ......update age) | Error: ", err)
	}

	fmt.Println("Command: db.customer.updateOne(......find by name, ......update age) | Result: ", updateResult)

	// Attempt to update a non-existing document, no document matched, nothing changed
	filterByNonExistingName := bson.M{"name": "Non Existing Name"}
	updateNonExisting := bson.M{"$set": bson.M{"age": 100}}
	updateResult, err = customerCollection.UpdateOne(ctx, filterByNonExistingName, updateNonExisting)
	if err != nil {
		fmt.Println("Command: db.customer.updateOne(......find by non existing name, ......update age) | Error: ", err)
	}

	fmt.Println("Command: db.customer.updateOne(......find by non existing name, ......update age) | Result: ", updateResult)

	// Attempt to update a non-existing document with upsert, creating a new document
	opts := options.Update().SetUpsert(true)
	updateResult, err = customerCollection.UpdateOne(ctx, filterByNonExistingName, updateNonExisting, opts)
	if err != nil {
		fmt.Println("Command: db.customer.updateOne(......find by non existing name, ......update age, .....provide upsert option) | Error: ", err)
	}

	fmt.Println("Command: db.customer.updateOne(......find by non existing name, ......update age, .....provide upsert option) | Result: ", updateResult)

	// Check the upserted document
	var nonExistingCustomer bson.M
	err = customerCollection.FindOne(ctx, filterByNonExistingName).Decode(&nonExistingCustomer)
	if err != nil {
		fmt.Println("Command: db.customer.find({ \"name\":  \"Non Existing Name\" }) | Error: ", err)
	}

	fmt.Println("Command: db.customer.find({ \"name\":  \"Non Existing Name\" }) | Result: ", nonExistingCustomer)

}
